src/NUnitFramework/framework/Constraints/Comparers/EquatablesComparer.cs

```diff
@@ -61,35 +61,53 @@ internal EquatablesComparer(NUnitEqualityComparer equalityComparer)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
-            var pair = new KeyValuePair<Type, MethodInfo>();
+            var mostDerived = default(EquatableMethodImpl);
 
-            foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
-                if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
-                        pair = xEquatableArgument;
+            foreach (var implementation in GetEquatableImplementations(first))
+            {
+                if (implementation.Argument.IsAssignableFrom(second))
+                {
+                    if (mostDerived.Argument == null || mostDerived.Argument.IsAssignableFrom(implementation.Argument))
+                        mostDerived = implementation;
+                }
+            }
 
-            return pair.Value;
+            return mostDerived.Method;
         }
 
-        private static IList<KeyValuePair<Type, MethodInfo>> GetEquatableGenericArguments(Type type)
+        private static EquatableMethodImpl[] GetEquatableImplementations(Type type)
         {
-            var genericArgs = new List<KeyValuePair<Type, MethodInfo>>();
+            static bool IsIEquatableOfT(Type t, object filter) => t.IsGenericType && t.GetGenericTypeDefinition().Equals(typeof(IEquatable<>));
+            
+            var interfaces = type.FindInterfaces((t,f) => IsIEquatableOfT(t,f), string.Empty);
+            var implementations = new EquatableMethodImpl[interfaces.Length];
 
-            foreach (Type @interface in type.GetInterfaces())
+            for(var i = 0; i < interfaces.Length; i++)
```

Comment:
Generally I'd prefer to see a Linq `Select(i => new EquatableMethodImpl(...)).ToArray()` but this is okay, it wasn't Linq before 😉